asciidoc/client-applications.adoc

```diff
@@ -63,87 +57,67 @@ include::{javascript-examples}/examples.test.js[tags=driver-lifecycle]
-In such a setup, a routing driver can include a preferred routing context via the query part of the `bolt+routing` URI.
+DNS resolution is the default, and always-available option.
+As it is possible to configure DNS to resolve a single host name down to multiple IP addresses, this can be used to expose all core server IP addresses under a single host name.
 
-In the standard Neo4j configuration, routing contexts are defined on the server side by means of _server policies_.
-Thus the driver communicates the routing context to the cluster in the form of a server policy.
-It then obtains refined routing information back from the cluster, based on the server policy.
+image::dns-resolution.png[title="Initial address resolution over DNS", role="middle"]
 
-The address in the URI of a routing driver with routing context must be that of a Core Server.
+==== Custom middleware
 
-.Configure a routing driver with routing context
-====
-This example will assume that Neo4j has been configured for server policies as described in <<operations-manual#multi-dc-load-balancing, Neo4j Operations Manual -> Load balancing for multi-data center systems>>.
-In particular, a server policy called `europe` has been defined.
-Additionally, we have a server `neo01.graph.example.com` to which we wish to direct the driver.
-
-This URI will use the server policy `europe`:
-
-`bolt+routing://neo01.graph.example.com?policy=europe`
-====
-
-[NOTE]
-.Server-side configuration to enable Routing drivers with routing context
-====
-A prerequisite for using a Routing driver with routing context is that the Neo4j database is operated on a <<operations-manual#clustering, Causal Cluster>> with the <<operations-manual#multi-dc, Multi-data center licensing option>> enabled.
-Additionally, the routing contexts must be defined within the cluster as routing _policies_.
-For details on how to configure multi-data center routing policies for a Causal Cluster, please refer to <<operations-manual#multi-dc-load-balancing, Operations Manual -> Causal Clustering>>.
-====
+Middleware, such as a load balancer, can be used to group the core servers under a single public address.
 
+image::custom-middleware.png[title="Initial address resolution using custom middleware"]
 
-=== Cluster bootstrap address resolution
+[[driver-resolver-function]]
+==== Resolver function
 
-Drivers have the ability to intercept cluster bootstrap address resolution.
-This is generally performed by a routing driver on construction or after complete contact has been lost with a cluster.
-This hook can be useful to augment regular DNS resolution and can be used to initialize a routing driver with multiple server addresses.
+Neo4j drivers also present an address resolution intercept hook called the resolver function.
+This takes the form of a callback function that accepts a single input address and yields or returns multiple output addresses.
```

Comment:
Suggestion:
This takes the form of a callback function that accepts a single input address and returns multiple output addresses.

Reply:
Absolutely. Fixed.